google/cloud/internal/populate_common_options.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/populate_common_options.h"
+#include "google/cloud/common_options.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/internal/user_agent_prefix.h"
+#include "absl/strings/str_split.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+std::set<std::string> DefaultTracingComponents() {
+  auto tracing =
+      google::cloud::internal::GetEnv("GOOGLE_CLOUD_CPP_ENABLE_TRACING");
+  if (!tracing.has_value()) return {};
+  return absl::StrSplit(*tracing, ',');
+}
+
+Options PopulateCommonOptions(Options opts, std::string const& endpoint_env_var,
+                              std::string const& emulator_env_var,
+                              std::string default_endpoint) {
+  if (!opts.has<EndpointOption>()) {
+    if (endpoint_env_var.empty()) {
+      opts.set<EndpointOption>(std::move(default_endpoint));
+    } else {
+      auto e = GetEnv(endpoint_env_var.c_str());
+      if (e && !e->empty()) {
+        opts.set<EndpointOption>(*std::move(e));
+      } else {
+        opts.set<EndpointOption>(default_endpoint);
+      }
+    }
+  }
+  if (!emulator_env_var.empty()) {
+    auto e = GetEnv(emulator_env_var.c_str());
+    if (e && !e->empty()) {
+      opts.set<EndpointOption>(*std::move(e));
+    }
+  }
```

Comment:
I think this is simpler, now that we are not setting the `GrpcCredentialOption` in here.
```cc
  if (!emulator_env_var.empty()) {
    auto e = GetEnv(emulator_env_var.c_str());
    if (e && !e->empty()) {
      opts.set<EndpointOption>(*std::move(e));
    }
  } else {
    if (!opts.has<EndpointOption>()) {
      opts.set<EndpointOption>(std::move(default_endpoint));
    }
  }
```

Reply:
The second `if()` would have no effect, we already assigned an `EndpointOption` in the `if()` starting at line 35.

Changed the code again to (maybe) make it easier to follow.

Reply:
I am sorry. I gave you bad code. (notably it never referenced `endpoint_env_var`)

Here is code that actually passes your tests. I claim that it is easier to follow:

```cc
  if (!emulator_env_var.empty()) {
    auto e = internal::GetEnv(emulator_env_var.c_str());
    if (e && !e->empty()) {
      opts.set<EndpointOption>(*std::move(e));
    }
  }
  if (!opts.has<EndpointOption>()) {
    auto e = internal::GetEnv(endpoint_env_var.c_str());
    if (e && !e->empty()) {
      opts.set<EndpointOption>(*std::move(e));
    } else {
      opts.set<EndpointOption>(std::move(default_endpoint));
    }
  }
```

Reply:
The tests do not test that setting `EndpointOption == ""` should be overwritten btw. I don't think this is a problem, the same way I didn't think we should have to test for an empty `std::function`. But just an FYI